kpatch-build/lookup.c

```diff
@@ -48,12 +48,77 @@ struct lookup_table {
 	int fd, nr;
 	Elf *elf;
 	struct symbol *syms;
+	struct symbol *local_syms;
 };
 
 #define for_each_symbol(ndx, iter, table) \
 	for (ndx = 0, iter = table->syms; ndx < table->nr; ndx++, iter++)
 
-struct lookup_table *lookup_open(char *path)
+static void resolve_dup_file_symbol(struct lookup_table *table, char *hint,
+				    struct sym_compare_type *locals)
+{
+	struct symbol *sym, *prev;
+	int i, in_file = 0;
+	int multi_base = 0;
+	struct sym_compare_type *local_index;
+
+	for_each_symbol(i, sym, table) {
+		if (sym->type == STT_FILE) {
+			if (in_file) {
+				if (table->local_syms) {
+					multi_base = 1;
+					in_file = 0;
+					break;
+				}
+				table->local_syms = prev;
+			}
+
+			if (!strcmp(hint, sym->name)) {
+				in_file = 1;
+				prev = sym;
+				local_index = locals;
+			}
+			else
+				in_file = 0;
+
+			continue;
+		}
+
+		if (!in_file)
+			continue;
+		if (sym->bind != STB_LOCAL || (sym->type != STT_FUNC && sym->type != STT_OBJECT))
+			continue;
+
+		for ( ; local_index->name; local_index++)
+			if (local_index->type == sym->type &&
+			    !strcmp(local_index->name, sym->name))
+				break;
```

Comment:
Yes, that's what I did before. But later we encountered a fail when find_local_syms().

Reproducer:
kernel: 3.10.34
gcc: gcc version 4.8.5 20150623 (Red Hat 4.8.5-4) (GCC)
file: fs/hugetlbfs/inode.c

local symbols in vmlinux of this inode.c :
 23718: ffffffff8126da80    16 FUNC    LOCAL  DEFAULT    1 hugetlbfs_write_begin
 23719: ffffffff8126da90   251 FUNC    LOCAL  DEFAULT    1 hugetlbfs_file_mmap
 23720: ffffffff8126db90    44 FUNC    LOCAL  DEFAULT    1 hugetlbfs_set_page_dirty
 23721: ffffffff8126dbc0   263 FUNC    LOCAL  DEFAULT    1 hugetlbfs_read_actor
 23722: ffffffff8126dcd0   480 FUNC    LOCAL  DEFAULT    1 hugetlbfs_read
 23723: ffffffff8126deb0   189 FUNC    LOCAL  DEFAULT    1 hugetlbfs_statfs
 23724: ffffffff8126df70    11 FUNC    LOCAL  DEFAULT    1 hugetlbfs_write_end
 23725: ffffffff8126df80    57 FUNC    LOCAL  DEFAULT    1 hugetlbfs_migrate_page
 23726: ffffffff81c2fec9   114 FUNC    LOCAL  DEFAULT   26 exit_hugetlbfs_fs
 23727: ffffffff81f2f950    16 OBJECT  LOCAL  DEFAULT   29 hugetlbfs_vfsmount
 23728: ffffffff81f2f960     8 OBJECT  LOCAL  DEFAULT   29 hugetlbfs_inode_cachep
 23729: ffffffff819fe700    56 OBJECT  LOCAL  DEFAULT   15 hugetlbfs_fs_type
 23730: ffffffff819fe740   680 OBJECT  LOCAL  DEFAULT   15 hugetlbfs_backing_dev_inf
 23731: ffffffff8126dfc0    26 FUNC    LOCAL  DEFAULT    1 hugetlbfs_mount
 23732: ffffffff8126e100  1170 FUNC    LOCAL  DEFAULT    1 hugetlbfs_fill_super
 23733: ffffffff8126dfe0    58 FUNC    LOCAL  DEFAULT    1 hugetlbfs_put_super
 23734: ffffffff8126e020    30 FUNC    LOCAL  DEFAULT    1 hugetlbfs_i_callback
 23735: ffffffff8126e040    20 FUNC    LOCAL  DEFAULT    1 init_once
 23736: ffffffff8126e060   158 FUNC    LOCAL  DEFAULT    1 hugetlbfs_alloc_inode
 23737: ffffffff81ae8546   368 FUNC    LOCAL  DEFAULT   18 init_hugetlbfs_fs
 23738: ffffffff8169db40   112 OBJECT  LOCAL  DEFAULT    4 tokens
 23739: ffffffff8169dee0   176 OBJECT  LOCAL  DEFAULT    4 hugetlbfs_ops
 23740: ffffffff8169dc80   192 OBJECT  LOCAL  DEFAULT    4 hugetlbfs_dir_inode_opera
 23741: ffffffff8126e5a0   581 FUNC    LOCAL  DEFAULT    1 hugetlb_get_unmapped_area
 23742: ffffffff8126e7f0   404 FUNC    LOCAL  DEFAULT    1 hugetlbfs_get_inode
 23743: ffffffff8169de20   168 OBJECT  LOCAL  DEFAULT    4 hugetlbfs_aops
 23744: ffffffff8169dbc0   192 OBJECT  LOCAL  DEFAULT    4 hugetlbfs_inode_operation
 23745: ffffffff8126e990   119 FUNC    LOCAL  DEFAULT    1 hugetlbfs_mknod
 23746: ffffffff8126ea10    55 FUNC    LOCAL  DEFAULT    1 hugetlbfs_mkdir
 23747: ffffffff8126ea50    26 FUNC    LOCAL  DEFAULT    1 hugetlbfs_create
 23748: ffffffff8126ea70   168 FUNC    LOCAL  DEFAULT    1 hugetlbfs_symlink
 23749: ffffffff8126eb20   482 FUNC    LOCAL  DEFAULT    1 truncate_hugepages
 23750: ffffffff8126ed10   347 FUNC    LOCAL  DEFAULT    1 hugetlbfs_setattr
 23751: ffffffff8126ee70    52 FUNC    LOCAL  DEFAULT    1 hugetlbfs_evict_inode
 23752: ffffffff8126eeb0    98 FUNC    LOCAL  DEFAULT    1 hugetlbfs_destroy_inode
 23753: ffffffff819fe680   128 OBJECT  LOCAL  DEFAULT   15 anon_ops
 23754: ffffffff81f2f944     1 OBJECT  LOCAL  DEFAULT   29 __print_once.30939
 23755: ffffffff81c18d58     8 OBJECT  LOCAL  DEFAULT   19 __initcall_init_hugetlbfs

local symbols of inode.o :
     5: 0000000000000000    16 FUNC    LOCAL  DEFAULT    1 hugetlbfs_write_begin
     6: 0000000000000010   251 FUNC    LOCAL  DEFAULT    1 hugetlbfs_file_mmap
     7: 0000000000000110    44 FUNC    LOCAL  DEFAULT    1 hugetlbfs_set_page_dirty
     9: 0000000000000140   263 FUNC    LOCAL  DEFAULT    1 hugetlbfs_read_actor
    13: 0000000000000250   480 FUNC    LOCAL  DEFAULT    1 hugetlbfs_read
    14: 0000000000000430   189 FUNC    LOCAL  DEFAULT    1 hugetlbfs_statfs
    16: 00000000000004f0    11 FUNC    LOCAL  DEFAULT    1 hugetlbfs_write_end
    18: 0000000000000500    57 FUNC    LOCAL  DEFAULT    1 hugetlbfs_migrate_page
    20: 0000000000000000   114 FUNC    LOCAL  DEFAULT   16 exit_hugetlbfs_fs
    21: 0000000000000010    16 OBJECT  LOCAL  DEFAULT    5 hugetlbfs_vfsmount
    22: 0000000000000020     8 OBJECT  LOCAL  DEFAULT    5 hugetlbfs_inode_cachep
    23: 0000000000000080    56 OBJECT  LOCAL  DEFAULT    3 hugetlbfs_fs_type
    24: 00000000000000c0   680 OBJECT  LOCAL  DEFAULT    3 hugetlbfs_backing_dev_inf
    25: 0000000000000540    26 FUNC    LOCAL  DEFAULT    1 hugetlbfs_mount
    26: 0000000000000680  1170 FUNC    LOCAL  DEFAULT    1 hugetlbfs_fill_super
    27: 0000000000000560    58 FUNC    LOCAL  DEFAULT    1 hugetlbfs_put_super
    28: 00000000000005a0    30 FUNC    LOCAL  DEFAULT    1 hugetlbfs_i_callback
    29: 00000000000005c0    20 FUNC    LOCAL  DEFAULT    1 init_once
    30: 00000000000005e0   158 FUNC    LOCAL  DEFAULT    1 hugetlbfs_alloc_inode
    33: 0000000000000000   368 FUNC    LOCAL  DEFAULT   19 init_hugetlbfs_fs
    34: 0000000000000040   112 OBJECT  LOCAL  DEFAULT   21 tokens
    36: 00000000000003e0   176 OBJECT  LOCAL  DEFAULT   21 hugetlbfs_ops
    37: 0000000000000180   192 OBJECT  LOCAL  DEFAULT   21 hugetlbfs_dir_inode_opera
    38: 0000000000000b20   581 FUNC    LOCAL  DEFAULT    1 hugetlb_get_unmapped_area
    39: 0000000000000d70   404 FUNC    LOCAL  DEFAULT    1 hugetlbfs_get_inode
    40: 0000000000000320   168 OBJECT  LOCAL  DEFAULT   21 hugetlbfs_aops
    41: 00000000000000c0   192 OBJECT  LOCAL  DEFAULT   21 hugetlbfs_inode_operation
    42: 0000000000000f10   119 FUNC    LOCAL  DEFAULT    1 hugetlbfs_mknod
    43: 0000000000000f90    55 FUNC    LOCAL  DEFAULT    1 hugetlbfs_mkdir
    44: 0000000000000fd0    26 FUNC    LOCAL  DEFAULT    1 hugetlbfs_create
    45: 0000000000000ff0   168 FUNC    LOCAL  DEFAULT    1 hugetlbfs_symlink
    46: 00000000000010a0   482 FUNC    LOCAL  DEFAULT    1 truncate_hugepages
    47: 0000000000001290   347 FUNC    LOCAL  DEFAULT    1 hugetlbfs_setattr
    48: 00000000000013f0    52 FUNC    LOCAL  DEFAULT    1 hugetlbfs_evict_inode
    49: 0000000000001430    98 FUNC    LOCAL  DEFAULT    1 hugetlbfs_destroy_inode
    50: 0000000000000000   128 OBJECT  LOCAL  DEFAULT    3 anon_ops
    51: 0000000000000004     1 OBJECT  LOCAL  DEFAULT    5 __print_once.30939
    53: 0000000000000000     8 OBJECT  LOCAL  DEFAULT   23 __exitcall_exit_hugetlbfs
    55: 0000000000000000     8 OBJECT  LOCAL  DEFAULT   25 __initcall_init_hugetlbfs

We can see there is a difference: inode.o includes '__exitcall_exit_hugetlbfs' , but vmlinux doesn't.
So the one-to-one match will fail. I don't know why the Linker drop this local symbol from inode.o.